Protein motifs are conserved fragments occurred frequently in protein sequences. 

They have significant functions, such as active site of an enzyme. 

Search and clustering protein sequence motifs are computational intensive. 

Most existing methods are not fast enough to analyze large data sets for motif finding or achieve low accuracy for motif clustering. 

We present a new protein sequence motif finding and clustering algorithm, called HSEARCH. 

It converts fixed length protein sequences to data points in high dimensional space, and applies locality-sensitive hashing to fast search homologous protein sequences for a motif. 

HSEARCH is significantly faster than the brute force algorithm for protein motif finding and achieves high accuracy for protein motif clustering.

For more details, please read this paper https://arxiv.org/abs/1701.00452.
